File: AtCoder/abc/abc099/c.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int main(){
    cin.tie(nullptr);
    ios::sync_with_stdio(false);

    int n;
    cin >> n;
    vector<int> use;
    use.push_back(1);

    int base6 = 6;
    while(base6 <= n){
        use.push_back(base6);
        base6 *= 6;
    }

    int base9 = 9;
    while(base9 <= n){
        use.push_back(base9);
        base9 *= 9;
    }

    sort(use.begin(), use.end());
    
    const int INF = 1001001001;
    vector<int> dp(n+1, INF);
    dp[0] = 0;
    dp[1] = 1;
    for(int i = 2; i < n+1; ++i){
        for(int val: use){
            if(i-val >= 0) dp[i] = min(dp[i], dp[i-val] + 1);
        }
    }

    cout << dp[n] << endl;


    return 0;
}

File: AtCoder/abc/abc099/c.py
n = int(input())
INF = 1001001001
dp = [INF] * 110000

dp[1] = 1
base = 6
while(base < 110000):
    dp[base] = 1
    base *= 6
base = 9
while(base < 110000):
    dp[base] = 1
    base *= 9

for i in range(101000):
    
    dp[i+1] = min(dp[i+1], dp[i] + 1)

    base = 6
    while(i+base < 110000):
        dp[i+base] = min(dp[i+base], dp[i] + 1)
        base *= 6

    base = 9
    while(i+base < 110000):
        dp[i+base] = min(dp[i+base], dp[i] + 1)
        base *= 9

print(dp[n])
